build_runner/lib/src/entrypoint/test.dart

```diff
@@ -0,0 +1,154 @@
+    // arguments after the `--`.
+    if (argResults.rest.isNotEmpty) {
+      void throwUsageException() {
+        throw new UsageException(
+            'The `test` command does not support positional args before the, '
+            '`--` separator, which should separate build args from test args.',
+            usage);
+      }
+
+      var separatorPos = argResults.arguments.indexOf('--');
+      if (separatorPos < 0) {
+        throwUsageException();
+      }
+      var expectedRest = argResults.arguments.skip(separatorPos + 1).toList();
+      if (argResults.rest.length != expectedRest.length) {
+        throwUsageException();
+      }
+      for (var i = 0; i < argResults.rest.length; i++) {
+        if (expectedRest[i] != argResults.rest[i]) {
+          throwUsageException();
+        }
+      }
+    }
+
+    return new SharedOptions.fromParsedArgs(
+        argResults, ['test'], packageGraph.root.name, this);
+  }
+
+  @override
+  Future<int> run() async {
+    SharedOptions options;
+    // We always run our tests in a temp dir.
+    var tempPath = Directory.systemTemp
+        .createTempSync('build_runner_test')
+        .absolute
+        .uri
+        .toFilePath();
+    try {
+      _ensureBuildTestDependency(packageGraph);
+      options = readOptions();
+      var outputMap = options.outputMap ?? {};
+      outputMap.addAll({tempPath: null});
+      var result = await build(
+        builderApplications,
+        deleteFilesByDefault: options.deleteFilesByDefault,
+        enableLowResourcesMode: options.enableLowResourcesMode,
+        failOnSevere: options.failOnSevere,
+        configKey: options.configKey,
+        assumeTty: options.assumeTty,
+        outputMap: outputMap,
+        packageGraph: packageGraph,
+        trackPerformance: options.trackPerformance,
+        skipBuildScriptCheck: options.skipBuildScriptCheck,
+        verbose: options.verbose,
+        builderConfigOverrides: options.builderConfigOverrides,
+        isReleaseBuild: options.isReleaseBuild,
+        buildDirs: options.buildDirs,
+      );
+
+      if (result.status == BuildStatus.failure) {
+        stdout.writeln('Skipping tests due to build failure');
+        return result.failureType.exitCode;
+      }
+
+      var testExitCode = await _runTests(tempPath);
+      if (testExitCode != 0) {
+        // No need to log - should see failed tests in the console.
+        exitCode = testExitCode;
+      }
+      return testExitCode;
+    } on _BuildTestDependencyError catch (e) {
+      stdout.writeln(e);
+      return ExitCode.config.code;
+    } finally {
+      // Clean up the output dir.
+      await new Directory(tempPath).delete(recursive: true);
+    }
+  }
+
+  /// Runs tests using [precompiledPath] as the precompiled test directory.
+  Future<int> _runTests(String precompiledPath) async {
+    stdout.writeln('Running tests...\n');
+    var extraTestArgs = argResults.rest;
+    var testProcess = await Process.start(
+        pubBinary,
```

Comment:
FYI: you don't want to assume `pub` is in the path. Use tricks here to find the `pub` associated w/ the SDK invoking the parent.

Reply:
Sure but im not changing any code in this pr just reorganizing :D